Updating Demonology APL for more damage

smadam813 opened the pull request:
### Notes on Demonology changes I made:

**Why change Imploding at 6 Imps instead of 8?**
It sims better.  6 seems to be the magic number when I simmed Imploding 3 imps iterating up to 12 imps.

**Why did I add `-- grimoire_felguard` and `-- summon_vilefiend` to SummonTyrant() function?**
When actually playing and you are moving from pack to pack / get interrupted etc, often it will run TyrantPrep, but you won't actually use your tyrant.  Moving to next pack while not exiting combat has been enough time for those two cooldowns to come back up.  Both pets considerably boost your tyrants damage with consumption.  If you were to follow APL as written, it would cast both those pets after Tyrant was summoned instead of before in that situation.

**Why remove `-- demonic_strength,if=!talent.demonic_consumption.enabled` from TyrantPrep()?**
There is no reason to cast demonic strength before your Tyrant as the damage boost is huge after Tyrant.  This doesn't matter if you are running demonic consumption or not.  Also casting demonic strength uses a valuable GCD while using nether portal / performing consumption prep.

**What is the BuildShard() function and why is it needed?**
When running Soul Strike, this is a handy tool for building up your shards during Tyrant prep phase or any time you would normally use Shadow Bolt.  There is no downside to using Soul Strike on CD and it should be utilized as soon as it is able to be cast.  Changing this improved damage sims when talented into Soul Strike.  Much smoother openers as well within M+ where the environment is fast paced.

### Simulations for APL Comparisons

**Patchwork Single Target APL Comparisons:**
https://www.raidbots.com/simbot/report/mxcpMZ7uuu1TCvA14khr6S

**Patchwork 3 Target APL Comparisons:**
https://www.raidbots.com/simbot/report/88pBfKJhfjCVJm2mTUk33f

**Dungeon Slice APL Comparisons:**
https://www.raidbots.com/simbot/report/nJhntPeFXtFnPrr6bHngbL

smadam813 committed:
Updating Demonology APL for more damage

smadam813 commented:
There might have to be more work done around number of imps to implode on while using Grim-Veiled Spaulders.  6 imp implosions sim better for every fight not using that legendary.  When running Grim-Veiled Spaulders however, 8 imp implosions have the potential of producing better results.  (Example would be Heavy Movement 3 Target sim)  I will working on figuring out thresholds, however for most fights, 6 produces higher dps even when running Grim-Veiled Spaulders.

Azevara commented:
@smadam813 Since you have several changes that are seemingly unrelated, can you provide comparison reports for each set of changes individually?
Example, for just the GFG/VF changes, they are not actually used it seems: https://www.raidbots.com/simbot/report/5wQwDPk6Lmn4EXaiMFu3KT

smadam813 commented:
The GFG/VF change is 100% a Murphy's law type of change.  I don't expect it ever having an impact in simulations which do not replicate a real world environment such as player being silenced etc.  The scenario to trigger these changes would be:

Player Character didn't cast Vilefiend when in Tyrant_Prep due to it being on cooldown.  PC enters Summon_Tyrant function since TyrantReady variable is true.  PC is silenced.  Vilefiend comes off of cooldown and can now be cast.  PC should now cast Vilefiend to empower Tyrant, however because TyrantReady variable is true, Vilefiend will never be cast until TyrantReady variable is set to false.

This was 100% a needed change for the herorotation for example.  I can remove these two lines from the APL as they will most likely never even be hit, but they really aren't hurting anything either.

smadam813 commented:
To show removing Demonic Strength from Tyrant_Prep phase I will be using Sacrificed Souls Talent.  Nether portal produces quite different results each time I run it due to the types of demons the portal summons and how many procs of demonbolt you get during Nether Portal rotation.  Sometimes it was higher, sometimes it was lower.  I will post sim results below however when taking Sacrificed Souls talent.

Simulations with talents: 3103011 (Highest DPS Talents for Sacrificed Souls using T26 Warlock)

Single Target: https://www.raidbots.com/simbot/report/mnrzw1Qc9pJnAmxmhEZBHr
New APL = -1 DPS

Dungeon Slice: https://www.raidbots.com/simbot/report/hhtnCHyUEpzE7z2Y2fP5Y
New APL = +80 DPS

Hectic Add Cleave: https://www.raidbots.com/simbot/report/uvBB1AUBaBdn3cHFMpmCET
New APL = +120 DPS

As you can see, in AoE encounters the DPS is higher due to that first demonic strength being cast with the Tyrants damage empowerment.  Single target appears to be a wash simming at -1 dps from current APL with margin of error at ~3 DPS.  

I think the results of sims I posted before show the difference for soul strike change.

Side Note:  I think Simcraft casts Demonic Strength even if your felguard is currently blade storming.  In game, the very first thing your felguard does when you engage combat is use bladestorm.  You are not even able to cast demonic strength until your felguards bladestorm ends however the sim appears to be overlapping those abilities.

smadam813 commented:
Here are sims for just GFG/VF changes:

Single Target: https://www.raidbots.com/simbot/report/7vB3DYbwCguuB98JrAGi2B
New APL = -2 DPS

Dungeon Slice: https://www.raidbots.com/simbot/report/7quVVe4xVCkhJjFPxqauFT
New APL = +18 DPS

Hectic Add Cleave: https://www.raidbots.com/simbot/report/p6vpf9j3X64V1kKLmM94ub
New APL = +15 DPS

Actually does improve it, I thought these would be a wash for sims.

smadam813 commented:
Let me know if you need anymore data.  I would be happy to provide it however I will likely be traveling a bit this weekend so not sure I will be around.

smadam813 commented:
Any updates on this?

Azevara commented:
@smadam813 I've implemented a few of these suggestions in #5759 . This should cover Demonic Strength and Soul Strike changes. The GFG/VF changes I believe need a cleaner solution, as the situation can never occur except when a sim is run with something that forces movement and that movement occurs during that specific moment in time. Adding the actions to the other list simply papers over it in that case.

If you could provide a few example sims for the implosion multitarget changes, I can get that implemented as well. However, it may be the case that stats and/or talents shifts the optimal value, so I need to see the comparisons you were running in order to investigate that.

smadam813 commented:
Sure thing.  The biggest loss of DPS I found was definitely the lack of usage on Soul Strike and Demonic Strength being called in Tyrant Prep so thanks for merging those in.  The amount of imps to implode on will definitely change depending if you are running Forces of the Horned Nightmare vs anything else.  What I will do is take the Tier Warlock build.  Sim imps from 3 to 12 and share results.  I will then equip a different legendary such as Relic of Demonic Synergy as well as Wilfred's to see if we can get some averages and how much they change depending on which legendary is equipped.

smadam813 commented:
I also need to put in another pull request to change T26 Warlock to use talents 3103032 instead of 3303032.
Demonic Calling I am 99% sure beats out Doom.

![image](https://user-images.githubusercontent.com/4027686/105427962-5764de00-5c1c-11eb-8d0e-c74065e98be2.png)

smadam813 commented:
### Template Update:
**T26 Warlock Demo Build Updates (I will make a separate pull request for this, but this is my base for all Implosion Sims.)**
https://www.raidbots.com/simbot/report/gAHiNc755we2cqujPWUzff
- Switched talents from 3303032 to 3103032 (Doom -> Demonic Calling)
- Switched Chest Enchant from Eternal Stats to Eternal Insight
- Switched Weapon Enchant from Sinful Revelation to Celestial Guidance

Results: +114 DPS Patchwork Increase

### Implosion Sims:
**2 Target Patchwork Smart Sim:**
https://www.raidbots.com/simbot/report/jzhcyvcpF9T7gSWq2TcARx
- 6 Imps is within Margin of Error of 8 Imps for Horned Nightmare
- 6 Imps is best with Demonic Synergy and Wilfreds

**3 Target Patchwork Smart Sim:**
https://www.raidbots.com/simbot/report/f4E378JEkWj3rZHVAmrwAV
- 6 Imps is best with Demonic Synergy, Horned Nightmare, and Wilfreds

**5 Target Patchwork Smart Sim:**
https://www.raidbots.com/simbot/report/9Ds2UkxSBWLBV22aJrgcQk
- 6 Imps is best with Demonic Synergy, Horned Nightmare, and Wilfreds

**Hectic Add Cleave 1 Boss 10,000 Iterations:**
https://www.raidbots.com/simbot/report/4NWGA8GwGBR8WuZ4wfxepb
- 5 and 6 Imps are within Margin of Error of each other using Demonic Synergy and Horned Nightmare
- 6 Imps is best with Wilfreds

**My Conclusion:**
6 Imps for All the Things

smadam813 commented:
### Implosive Potential Sims:
**2 Target Patchwork Smart Sim:**
https://www.raidbots.com/simbot/report/n6pRxDFbtNQf9c6Udt1n25
- 6 Imps

**3 Target Patchwork Smart Sim:**
https://www.raidbots.com/simbot/report/combuEgnYjhy6WL1jxuECX
- 6 Imps

**5 Target Patchwork Smart Sim:**
https://www.raidbots.com/simbot/report/3Tqu95F5rG5p1m9Wf5tBU1
- 6 Imps (Whopping 1000 DPS increase over 8 Imps)

**Hectic Add Cleave 1 Boss 10,000 Iterations:**
https://www.raidbots.com/simbot/report/ncDb3h5nwpJcChWM7jXcTW
- 6 Imps

### Overall Conclusion Including Implosive Potential:
With how Implosion is currently implemented in the APL, imploding at **6 Imps** seems to be the best for every Demonology Legendary worth running.

smadam813 commented:
I would like to close out this PR if no actions are going to be taken.  I put a lot of work into simming everything out in order to show the differences.  Is there any update on this?

seanpeters86 commented:
@smadam813 can you resolve merge conflicts please?

smadam813 commented:
I will soon as I am able.

smadam813 committed:
Removing Build Shard Changes

smadam813 commented:
I fixed the merge conflicts and removed the build shard changes just leaving implosion adjustments from 8 imps to 6 imps + grimoire_felguard and summon_vilefiend changes.  I think there have been enough sims provided in this merge request to adequately prove these are improvements.  If there is still doubt let's just close this, it's been far too long and I don't have the energy to prove it more than I have.

smadam813 committed:
Forgot to add back Shadow Bolt for Summon Tyrant

smadam813 commented:
Thank you!
